src/experimental/queue.rs

```diff
@@ -0,0 +1,372 @@
+                        let c = Connection {
+                            query_ch: query_req_ch_snd,
+                            quit_ch: query_quit_ch_snd,
+                        };
+                        conns.insert(req.db_name.clone(), c);
+                        // TODO do something real to open the db.
+                        let store = Store {
+                            map: Box::new(HashMap::new()),
+                        };
+                        spawn_local(connection_loop(query_req_ch_rec, query_quit_ch_rec, store));
+                        req.response_sender.send("ok".into()).await;
+                        continue;
+                    }
+                    "close" => {
+                        let msg = if req.db_name.is_empty() {
+                            "db_name must be non-empty"
+                        } else {
+                            match conns.get(&req.db_name[..]) {
+                                Some(c) => {
+                                    c.quit_ch.send(()).await;
+                                    // TODO maybe wait for the quit to complete?
+                                    conns.remove(&req.db_name[..]);
+                                    "ok"
+                                }
+                                None => "db not open",
+                            }
+                        };
+                        req.response_sender.send(msg.into()).await;
+                        continue;
+                    }
+                    _ => match conns.get(&req.db_name[..]) {
+                        Some(c) => {
+                            c.query_ch.send(req).await;
+                        }
+                        None => {
+                            req.response_sender.send("no such db".into()).await;
+                        }
+                    },
+                }
+            }
+        }
+    }
+}
+
+// One per connection / database. Handles requests for this db.
+async fn connection_loop(req_ch_rec: Receiver<Request>, quit_ch_rec: Receiver<()>, mut store: Store) {
+    // We buffer Requests to the db so we can allow Requests to the current
+    // write tx (if any) to jump the queue.
+    let mut queue = VecDeque::new();
+
+    // If write_tx is None we are executing reads.
+    let mut write_tx: Option<WriteTx> = None;
+    let mut next_tx_id = 0u64;
+
+    loop {
+        // Check if we have received the quit signal.
+        match quit_ch_rec.try_recv() {
+            Err(TryRecvError::Empty) => {}
+            Err(TryRecvError::Disconnected) => {
+                // TODO log an error
+                return;
+            }
+            Ok(_) => return,
+        }
+
+        // We want to wait asynchronously for a new request (and not spin) if
+        // we have nothing to do. We have nothing to do if:
+        //   A) the queue is empty or
+        //   B) we have an open write tx but zero Requests for it in the queue
```

Comment:
I do not think that this is quite correct. If a read transaction is open while a write transactions starts, that read transaction will stall until the write transaction completes. Worse, the read transaction will read an inconsistent view of the world -- part before the write tx and part after.

I think that we need to do something more like I proposed in the sketch -- we keep track of open read transactions too -- we can either be in a mode where we're dispatching to n open read tx, or a mode where we are dispatching to one open write tx. Before we start the write tx, all open read tx must complete, just as you have us waiting for the write tx now.

Reply:
right. same technique can work for read tx.